File: src/ackward/logging/Handler.cpp
#include <ackward/logging/Handler.hpp>

#include <boost/python.hpp>

#include <ackward/logging/Filter.hpp>
#include <ackward/logging/Formatter.hpp>
#include <ackward/logging/Module.hpp>

using namespace boost::python;

namespace ackward { namespace logging 
{

Handler::Handler(boost::python::object h) :
    Object (h)
{}

void Handler::setLevel(Level l) 
{
    obj().attr("setLevel")(l);
}

void Handler::setFormatter(Formatter f) 
{
    obj().attr("setFormatter")(f);
}

void Handler::addFilter(Filter f) 
{
    obj().attr("addFilter")(f);
}
    
void Handler::removeFilter(Filter f) 
{
    obj().attr("removeFilter")(f);
}

void Handler::flush() const 
{
    obj().attr("flush")();
}

void Handler::close() 
{
    obj().attr("close")();
}

void Handler::emit(const LogRecord& r) const
{
    obj().attr("emit")(r);
}

StreamHandler::StreamHandler() :
    Handler ( module().attr("StreamHandler")() )
{}

StreamHandler::StreamHandler(boost::python::object stream) :
    Handler ( module().attr("StreamHandler")(stream) )
{}

FileHandler::FileHandler(const std::wstring& filename,
                         const std::string& mode, 
                         const std::string& encoding,
                         bool delay) :
    Handler (
        module().attr("FileHandler")(
            filename,
            mode,
            encoding.empty() 
            ? boost::python::object() 
            : boost::python::str(encoding),
            delay))
{}

WatchedFileHandler::WatchedFileHandler(const std::wstring& filename,
                                       const std::string& mode, 
                                       const std::string& encoding,
                                       bool delay) :
    Handler(
        import("logging.handlers").attr("WatchedFileHandler")(
            filename,
            mode,
            encoding.empty() 
            ? boost::python::object()
            : boost::python::str(encoding),
            delay))
{}

SocketHandler::SocketHandler(const std::wstring& host,
                             unsigned int port) :
    Handler(
        import("logging.handlers").attr("SocketHandler")(
            host, port))
{}

}}

File: src/ackward/logging/Handler.py
from ackward import (Class,
                     method,
                     Method,
                     Namespace,
                     TranslationUnit)

def tunit():
    return TranslationUnit(
        guard='INCLUDE_ACKWARD_LOGGING_HANDLER_HPP',
        forward_declarations=[
            ('ackward', 'logging', 'class Formatter'),
            ('ackward', 'logging', 'class Filter'),
            ('ackward', 'logging', 'class LogRecord'),
            ],
        header_includes=[
            ('ackward', 'core', 'Object.hpp'),
            ('ackward', 'logging', 'Types.hpp'),
            ],
        impl_includes=[
            ('ackward', 'logging', 'Formatter.hpp'),
            ('ackward', 'logging', 'Filter.hpp'),
            ('ackward', 'logging', 'Handler.hpp'),
            ('ackward', 'logging', 'LogRecord.hpp'),
            ])

def methods(parent):
    m = [
        ('void setLevel(Level level)',
         'Set the logging level of this handler.'),
        ('void setFormatter(Formatter f)',
         'Set the formatter for this handler.'),
        ('void addFilter(Filter f)',
         'Add the specified filter to this handler.'),
        ('void removeFilter(Filter f)',
         'Remove the specified filter from this handler.'),
        ('void flush() const',
         'Ensure all logging output has been flushed.'),
        ('void close()',
         'Tidy up any resources used by the handler.')
        ]
    list(map(lambda m: method(m[0], parent=parent, doc=m[1]), m))

    Method(
        name='emit',
        signature=[('LogRecord', 'r')],
        const=True,
        virtual=True,
        parent=parent,
        doc='Do whatever it takes to actually log the specified logging record.')

def definition(env):
    t = tunit()
    ns = Namespace('ackward', 'logging', parent=t)
    cls = Class(name='Handler',
                wrapped_class='logging.Handler',
                parent=ns,
                doc='Handler instances dispatch logging events to specific destinations.')
    methods(cls)

    return t


